sdk/nodejs/x/automation/localWorkspace.ts

```diff
@@ -0,0 +1,357 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as fs from "fs";
+import * as os from "os";
+import * as upath from "upath";
+
+import { CommandResult, runPulumiCmd } from "./cmd";
+import { ConfigMap, ConfigValue } from "./config";
+import { ProjectSettings } from "./projectSettings";
+import { Stack } from "./stack";
+import { StackSettings } from "./stackSettings";
+import { PulumiFn, StackSummary, Workspace } from "./workspace";
+
+export class LocalWorkspace implements Workspace {
+    ready: Promise<any[]>;
+    private workDir: string;
+    private pulumiHome?: string;
+    private program?: PulumiFn;
+    private envVars: { [key: string]: string };
+    private secretsProvider?: string;
+    public static async NewStackLocalSource(
```

Comment:
Should we use overloads instead of `LocalSource` vs `InlineSource` suffixes? Something like (this also removes the `upsert` variant in favor of a `create` arg on select; see comment on `Stack` for more details on this suggestion):

```typescript
export interface LocalProgramArgs {
    stackName: string;
    workDir: string;
}

export interface InlineProgramArgs {
    stackName: string;
    projectName: string;
    program: PulumiFn;
}

export class LocalWorkspace implements Workspace {
    public static createStack(args: LocalProgramArgs, opts?: LocalWorkspaceOpts): Promise<Stack>;
    public static createStack(args: InlineProgramArgs, opts?: LocalWorkspaceOpts): Promise<Stack>;
    static async createStack(args: LocalProgramArgs | InlineProgramArgs, opts?: LocalWorkspaceOpts): Promise<Stack> {
        if (isLocalProgramArgs(args)) {
            // Local implementation...
        } else if (isInlineProgramArgs(args)) {
            // Inline implementation...
        } else {
            throw new Error(`Unexpected args: ${args}`);
        }
    }

    public static selectStack(args: LocalProgramArgs, create?: boolean, opts?: LocalWorkspaceOpts): Promise<Stack>;
    public static selectStack(args: InlineProgramArgs, create?: boolean, opts?: LocalWorkspaceOpts): Promise<Stack>;
    static async selectStack(args: LocalProgramArgs | InlineProgramArgs, create = false, opts?: LocalWorkspaceOpts): Promise<Stack> {
        if (isLocalProgramArgs(args)) {
            // Local implementation...
        } else if (isInlineProgramArgs(args)) {
            // Inline implementation...
        } else {
            throw new Error(`Unexpected args: ${args}`);
        }
    }
}

function isLocalProgramArgs(args: LocalProgramArgs | InlineProgramArgs): args is LocalProgramArgs {
    return (args as LocalProgramArgs).workDir !== undefined;
}

function isInlineProgramArgs(args: LocalProgramArgs | InlineProgramArgs): args is InlineProgramArgs {
    return (args as InlineProgramArgs).projectName !== undefined &&
        (args as InlineProgramArgs).program !== undefined;
}
```

And usage:

```typescript
const program = async () => {
    console.log("hello");
}

const stack = await LocalWorkspace.createStack({
    stackName: "foo",
    projectName: "foo",
    program,
})
```

Reply:
Updated the usage accordingly. This will be nice as there's a third set of args that will eventually be added (RemoteProgramArgs).